DESIGN-DECISIONS.md

```diff
@@ -0,0 +1,342 @@
+`AllocatedByPostgres`, then it is left to Postgres to free the backing pointer whenever it thinks it should (typically 
+when its owning `MemoryContext` is deleted).  Only if it's thought to be `AllocatedByRust` is the backing pointer is 
+freed (via `pfree()`) when the `PgBox` instance is dropped.
+
+The `WhoAllocated` trait is a Zero Sized Type (ZST) and incurs no runtime overhead in terms of memory or branching -- 
+the compiler takes care of the implementation details through monomorphization.  Additionally, `PgBox` itself is only
+the size of a pointer.
+
+Functions such as `::alloc<T>()` exist (which delegate to `palloc()`) to create a new `PgBox` wrapping a pointer to a
+specific type.
+
+The `::into_pg(self)` function will relinquish `PgBox`'s ownership of the backing pointer and simply return it, which in
+turn passes responsibility of freeing the pointer to Postgres when it sees fit.
+
+`PgBox` also provides `Deref` and `DerefMut` implementations which will, as a safety measure, raise a Postgres ERROR 
+if the backing pointer is NULL.  These exist as a convenience for slightly more fluent code.
+
+(as an aside, the `WhoAllocated` trait which determines much of `PgBox`'s behavior is poorly named.  It should be
+thought of as "WhoDrops", as it doesn't matter if the user calls `palloc()` directly and provides the pointer, if
+`PgBox` calls `palloc()` itself, or if it consumes a pointer provided by Postgres.  The decision it is making is if the 
+backing pointer should be freed following Rust's normal drop semantics or not)
+
+### The Path to a Longer Life
+
+Common concerns around "who allocated what and how long does it live" are typically centered around Datums (see
+below), passing pointers to internal Postgres functions, and holding on to pointers (regardless of source) for
+longer than the current Rust scope or life of `CurrentMemoryContext`.
+
+For the later case where Rust-allocated bytes need to live longer than the compiler will allow, `PgMemoryContexts.leak_and_drop_on_delete(T)`
+allows the thing to be leaked (returned to the caller as `*mut T`) and later dropped when Postgres decides to delete 
+(free) the `MemoryContext` to which it was leaked.  The returned pointer can be stashed somewhere and later 
+re-instantiated via `PgBox::from_pg()`.
+
+## Thread Safety
+
+As it relates to any Postgres thing (calling a function, allocated memory, anything at all from `pgx::*::*`), **there is 
+none!**  Postgres itself is purposely not thread safe and neither is pgx.
+
+It's difficult (right now) to ask the Rust compiler to ensure code is not using pgx from a thread.  We have been 
+discussing some ideas.  Until then, there's one simple rule:
+
+- Do **not** use anything from the `pgx` crate or any Postgres allocated memory from a thread
+
+You might be really smart and think you can sneak by doing something, but you're probably wrong.  Even if you're right,
```

Comment:
It would not play well with our `#[pg_guard]` machinery even if you did -- that code writes to globals without synchronization.

Reply:
Well, I didn't want to give examples, but I was thinking someone might think they're smart enough to use something like Postgres' `pub static mut InterruptPending: bool;` to see if they ought to coax a thread into terminating.  Like, maaaaaaybe that'd work, but yeah, I ain't wanting to suggest it.

Reply:
It would probably not work on ARM, unless there's some additional fencing.